### PR TITLE
Add dry run

### PR DESCRIPTION
**Pull request by stefan-mysten**
Note: will need to add tests later once we fix the transaction bcs issue (intent is part of bcs).

### CODE REVIEW

File: crates/sui-graphql-client/src/lib.rs [hunk 1]
```diff
@@ -602,6 +602,55 @@ impl Client {
...
+    ///
+    ///  - if present, txBytes is assumed to be a [`TransactionKind`],
+    ///  - if absent, txBytes is assumed to be [`Transaction`].
+    ///
+    /// `skipChecks` optional flag to disable the usual verification checks that prevent access to
+    /// objects that are owned by addresses other than the sender, and calling non-public,
+    /// non-entry functions, and some other checks. Defaults to false.
+    pub async fn dry_run(
+        &self,
+        tx_bytes: Vec<u8>,
```
Comment (bmwill) [inline crates/sui-graphql-client/src/lib.rs:619]: Same comment as in the other PR, this should be a &Transaction not bytes

Comment (stefan-mysten) [reply crates/sui-graphql-client/src/lib.rs]: Fixed!

File: crates/sui-graphql-client/src/lib.rs [hunk 2]
```diff
@@ -602,6 +602,57 @@ impl Client {
...
+    /// `txBytes` is either a TransactionData struct or a [`TransactionKind`] struct, BCS-encoded
+    /// and then Base64-encoded. The expected type is controlled by the presence or absence of
+    /// `txMeta`:
+    ///
+    ///  - if present, txBytes is assumed to be a [`TransactionKind`],
+    ///  - if absent, txBytes is assumed to be [`Transaction`].
+    ///
+    /// `skipChecks` optional flag to disable the usual verification checks that prevent access to
+    /// objects that are owned by addresses other than the sender, and calling non-public,
+    /// non-entry functions, and some other checks. Defaults to false.
```
Comment (bmwill) [inline crates/sui-graphql-client/src/lib.rs:616]: Do we need to have two dry_run apis? This doc comment doesn't match the function signature now that it takes &Transaction...

Comment (stefan-mysten) [reply crates/sui-graphql-client/src/lib.rs:616]: I don't follow the `two dry_run apis`? Were you referring to the usage of `txMeta` to determine the type of the transaction to dry run?

I assume we can always just use `Transaction`, and do not pass the `txMeta` argument. I am not sure what was the rationale behind having this two-version dry-run in GraphQL.

Comment (bmwill) [reply crates/sui-graphql-client/src/lib.rs]: your comment says that the metadata only is used if a TransactionKind is sent, so a good api would split these two cases into separate apis, a Kind with metadata and a Transaction only api

Comment (stefan-mysten) [reply crates/sui-graphql-client/src/lib.rs]: Ah I see what you mean. Thanks for that!

Comment (stefan-mysten) [reply crates/sui-graphql-client/src/lib.rs]: Refactored

File: crates/sui-graphql-client/src/lib.rs [hunk 3]
```diff
@@ -683,6 +688,82 @@ impl Client {
...
+    /// Dry run a [`Transaction`] and return the transaction.
+    ///
+    /// `skipChecks` optional flag disables the usual verification checks that prevent access to
+    /// objects that are owned by addresses other than the sender, and calling non-public,
+    /// non-entry functions, and some other checks. Defaults to false.
+    pub async fn dry_run_tx(
+        &self,
+        tx: &Transaction,
+        skip_checks: Option<bool>,
+    ) -> Result<Option<Transaction>, Error> {
```
Comment (bmwill) [inline crates/sui-graphql-client/src/lib.rs:704]: This return type doesn't make any sense. If i'm performing a dryrun i want to preview the effects of the transaction, not get back the exact txn that i sent

File: crates/sui-graphql-client/src/lib.rs [hunk 4]
```diff
@@ -683,6 +688,82 @@ impl Client {
...
+    /// objects that are owned by addresses other than the sender, and calling non-public,
+    /// non-entry functions, and some other checks. Defaults to false.
+    ///
+    /// `tx_meta` is the transaction metadata.
+    pub async fn dry_run_tx_kind(
+        &self,
+        tx_kind: &TransactionKind,
+        skip_checks: Option<bool>,
+        tx_meta: TransactionMetadata,
+    ) -> Result<Option<Transaction>, Error> {
```
Comment (bmwill) [inline crates/sui-graphql-client/src/lib.rs:723]: Same here

File: crates/sui-graphql-client/src/lib.rs (hunk 3 above)
Comment (stefan-mysten) [reply crates/sui-graphql-client/src/lib.rs:704]: 🤦  yes, you're right.

File: crates/sui-graphql-client/src/query_types/dry_run.rs [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+pub struct Input {
+    pub __typename: String,
+    pub ix: i32,
+}
+
+#[derive(cynic::QueryFragment, Debug)]
+#[cynic(schema = "rpc", graphql_type = "GasCoin")]
+pub struct GasCoin {
+    #[cynic(rename = "_")]
+    pub __underscore: Option<bool>,
```
Comment (stefan-mysten) [inline crates/sui-graphql-client/src/query_types/dry_run.rs:36]: The GraphQL schema has a "dummy" field.

File: crates/sui-graphql-client/src/query_types/dry_run.rs [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+#[derive(cynic::QueryFragment, Debug)]
+#[cynic(schema = "rpc", graphql_type = "DryRunResult")]
+pub struct DryRunResult {
+    pub error: Option<String>,
+    pub results: Option<Vec<DryRunEffect>>,
+}
```
[the remark this answers is not in the document]
Comment (stefan-mysten) [reply crates/sui-graphql-client/src/query_types/dry_run.rs:44]: TIL: Returning the transaction from the dry-run gives us the effects as well, so we could just build the `TransactionEffects` from bcs and return only the effects. I will do that!

File: crates/sui-graphql-client/src/query_types/dry_run.rs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+#[derive(cynic::InputObject, Debug)]
+#[cynic(schema = "rpc", graphql_type = "ObjectRef")]
+pub struct ObjectRef {
+    pub address: Address,
+    pub digest: String,
+    pub version: Uint53,
+}
+
+impl From<ObjectReference> for ObjectRef {
```
Comment (stefan-mysten) [inline crates/sui-graphql-client/src/query_types/dry_run.rs:50]: This is useful when passing the required transaction metadata object, which includes the gas objects that is a list of `ObjectRef`, so one can just do `.into()` on those `ObjectReference`.

File: crates/sui-graphql-client/src/query_types/dry_run.rs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+pub struct DryRunQuery {
+    #[arguments(txBytes: $tx_bytes, skipChecks: $skip_checks, txMeta: $tx_meta)]
+    pub dry_run_transaction_block: DryRunResult,
+}
+
+#[derive(cynic::QueryFragment, Debug)]
+#[cynic(schema = "rpc", graphql_type = "DryRunResult")]
+pub struct DryRunResult {
+    pub error: Option<String>,
+    pub transaction: Option<TransactionBlock>,
```
Comment (bmwill) [inline crates/sui-graphql-client/src/query_types/dry_run.rs:21]: What is a TransactionBlock? the definition here only has a single bcs field which doesn't indicate if it includes events, effects, ect

Comment (stefan-mysten) [reply crates/sui-graphql-client/src/query_types/dry_run.rs:21]: Ah yes,  thanks for catching this mistake.

GraphQL calls TransactionBlock the type that includes most data (gas, effects, signatures, kind, sender, digest, expiration). GraphQL dry run will return the transaction block with its effects. 
The bcs of TransactionBlock is the `SenderSignedData`, bcs serialized and Base64 encoded, so it does not contain the effects. I'd need to add the `effects` field on the `TransactionBlock` and use the bcs field to convert to `sui-sdk-types::types::TransactionEffects`.

Comment (stefan-mysten) [reply crates/sui-graphql-client/src/query_types/dry_run.rs:21]: @bmwill fixed it now.

**Pull request merged**